library/aes.c

```diff
@@ -702,6 +720,47 @@ int mbedtls_aes_setkey_dec( mbedtls_aes_context *ctx, const unsigned char *key,
 
     return( ret );
 }
+
+#if defined(MBEDTLS_CIPHER_MODE_XTS)
+static int mbedtls_aes_xts_setkey( mbedtls_aes_xts_context *ctx,
+                                   const unsigned char *key,
+                                   unsigned int keybits,
+                                   int (*setkey1)( mbedtls_aes_context *ctx,
+                                                   const unsigned char *key,
+                                                   unsigned int keybits ) )
+{
+    const unsigned int key1bits = keybits / 2;
+    const unsigned int key2bits = keybits / 2;
+    const unsigned int key1bytes = key1bits / 8;
+    const unsigned char *key1 = &key[0];
+    const unsigned char *key2 = &key[key1bytes];
+    int ret;
+
+    ret = mbedtls_aes_setkey_enc( &ctx->tweak, key2, key2bits );
+
+    if( ret )
+        return( ret );
+
+    return( setkey1( &ctx->crypt, key1, key1bits ) );
+}
+
+int mbedtls_aes_xts_setkey_enc( mbedtls_aes_xts_context *ctx,
+                                const unsigned char *key,
+                                unsigned int keybits)
+{
+    return( mbedtls_aes_xts_setkey( ctx, key, keybits,
+                                    mbedtls_aes_setkey_enc ) );
```

Comment:
There is a clear benefit to having function pointers here: maintainability (by not repeating one's self), readability (by making clear the exact differences between two implementations, not burying differences between two implementations in a pile of similar looking code).

That said, this is a security sensitive library, so security concerns will trump both maintainability and readability. I'll reconsider if it is possible to meet all three needs: security, maintainability, and readability.

Reply:
I've updated the code to try to balance security, maintainability, and readability.

Reply:
I disagree that there's a maintainability benefit to a function pointer here. The part that's repeated is small and could be contained in a single function so that each of `mbedtls_aes_xts_setkey_dec` and `mbedtls_aes_xts_setkey_enc` could be just `{ int ret = common_function(ctx,key,keybits); if (ret == 0) ret = variable_part(); return ret; }`